### PR TITLE
feat(seo): added seo-service

### PR DESCRIPTION
**Pull request by marcin-hoa**
Implements: KAP-9
Fixes: KAP-101

I tried to inlcude all tags from `yoast_head` returned by WP Rest API and from [ngx-seo.service.ts](https://github.com/avivharuzi/ngx-seo/blob/main/packages/ngx-seo/src/lib/ngx-seo.service.ts).

It became pretty big, but there are a lot of tags. 

Currently it is provided and injected in `RootShellComponent`. Let me know if you know a better way to do this. 

Also I fixed mobile navigation (KAP-101)

### CODE REVIEW

File: libs/blog/shared/feature-seo/src/lib/seo/seo.service.ts [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+
+    this._meta.removeTag('itemprop="description"'); // Because if we not remove the tag it will not be updated.
+
+    this._meta.updateTag(description);
+    this._meta.updateTag(ogDescription);
+    this._meta.updateTag(twitterDescription);
+    this._meta.updateTag(googleDescription);
+  }
+
+  private setMetaType(metaType: string): void {
```
Comment (jakNatonek) [inline libs/blog/shared/feature-seo/src/lib/seo/seo.service.ts:221]: i would reduce number of code lines with that func: 

```
 private updateTag(content: string, name: SeoMetaKeys): void {
    const meta: MetaDefinition = {
      name: name,
      content: content,
    };

    this._meta.updateTag(meta);
  }
```

where SeoMetaKeys

`export type SeoMetaKeys = keyof typeof SEO_META_KEYS;`

and 

`SEO_META_KEYS = {...} as const`

File: libs/blog/shared/feature-seo/src/lib/seo/seo.service.ts [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+    const ogType: MetaDefinition = {
+      property: SEO_META_KEYS.ogType,
+      content: metaType,
+    };
+
+    this._meta.updateTag(ogType);
+  }
+
+  private setMetaCanonical(canocical: string): void {
```
Comment (jakNatonek) [inline libs/blog/shared/feature-seo/src/lib/seo/seo.service.ts:230]: do we need func for that single meta updates? i would remove that and just invoke 

`this.updateTag`  (func from prev comment)

File: libs/blog/shared/feature-seo/src/lib/seo/seo.service.ts (hunk 1 above)
Comment (marcin-hoa) [reply libs/blog/shared/feature-seo/src/lib/seo/seo.service.ts]: Done :)

File: libs/blog-contracts/articles/src/lib/articles.ts [hunk 3]
```diff
@@ -30,6 +30,38 @@ export interface ArticlePreview {
...
+  article_published_time?: string;
+  article_modified_time?: string;
+  og_image?: {
+    width: number;
+    height: number;
+    url: string;
+    type: string;
+  }[];
+  twitter_card?: string;
+  twitter_misc?: object;
```
Comment (DDonochVA) [inline libs/blog-contracts/articles/src/lib/articles.ts:62]: avoid `object`
use `Record<string, string>`

File: libs/blog/articles/data-access/src/lib/state/article-details.store.ts [hunk 4]
```diff
@@ -46,12 +46,13 @@ export const ArticleDetailsStore = signalStore(
                     slug: slug,
                     fetchArticleDetailsCallState: { error },
                   }),
-                next: (articleDetails) =>
-                  patchState(store, {
+                next: (articleDetails) => {
```
Comment (DDonochVA) [inline libs/blog/articles/data-access/src/lib/state/article-details.store.ts:49]: is this change needed?

File: libs/blog/shell/feature-shell-web/src/lib/blog-shell.routes.ts [hunk 5]
```diff
@@ -21,13 +21,19 @@ export const blogShellRoutes: Route[] = [
         loadComponent: async () =>
           (await import('@angular-love/feature-about-us'))
             .FeatureAboutUsComponent,
+        data: {
+          seoTitle: 'About Us',
+        },
```
Comment (DDonochVA) [inline libs/blog/shell/feature-shell-web/src/lib/blog-shell.routes.ts:26]: all seo parameters should be wrapped within some object like "seo", to aggregate all seo-related stuff into one container
```suggestion
        data: {
          seo: {
            title: "About Us"
          }
        },
```

**Review by DDonochVA — COMMENTED** [review]
what about article page?

File: libs/blog/shared/feature-seo/src/lib/seo/seo.service.ts [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+  private readonly _activatedRoute = inject(ActivatedRoute);
+  private readonly _title = inject(Title);
+  private readonly _meta = inject(Meta);
+
+  constructor() {
+    this.subscribe();
+  }
+
+  private setGlobalSeo(seoData: ArticleSeoData | undefined): void {
+    this.updateTag(seoData?.og_locale || 'en_US', 'ogLocale');
```
Comment (DDonochVA) [inline libs/blog/shared/feature-seo/src/lib/seo/seo.service.ts:26]: create an injection token to accept default data, so it won't hold static values here.


```ts
export type SeoConfig = { };
export const SEO_CONFIG = new InjectionToken<SeoConfig>('SEO_CONFIG');
export const provideSeo = (config?: SeoConfig) => {
  return makeEnvironmentProvides([
    {
      provide: SEO_CONFIG,
      useValue: config,
    },
    SeoService
  ])
}
```

File: libs/blog/shared/feature-seo/src/lib/seo/seo.service.ts [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+  }
+
+  private setTitle(title: string): void {
+    this._title.setTitle(title);
+    this.updateTag(title, 'ogTitle');
+    this.updateTag(title, 'twitterTitle');
+    this.updateTag(title, 'name');
+  }
+
+  private setMeta(seoData: ArticleSeoData): void {
```
Comment (DDonochVA) [inline libs/blog/shared/feature-seo/src/lib/seo/seo.service.ts:55]: remove the `ArticleSeoData` dependency, create a different interface for that

File: libs/blog/shared/feature-seo/src/lib/seo/seo.service.ts [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+    this.setMetaDescription(
+      seoData?.description ||
+        'Angular.love - ciekawostki oraz rozwiązania dla średnio-zaawansowanych oraz zaawansowanych developerów Angulara.',
+    );
+
+    this.updateTag(seoData?.og_site_name || 'Angular.love', 'ogSiteName');
+  }
+
+  private setArticleSeo(seoData: ArticleSeoData | undefined): void {
```
Comment (DDonochVA) [inline libs/blog/shared/feature-seo/src/lib/seo/seo.service.ts:36]: this method can me removed if we expose `setTitle` and `setMeta` to public.

File: libs/blog/shared/feature-seo/src/lib/seo/seo.service.ts [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+
+    const params$ = routerNavigation$.pipe(
+      switchMap((route) => route?.paramMap),
+    );
+    const routerData$ = routerNavigation$.pipe(
+      switchMap((route) => route?.data),
+    );
+
+    combineLatest({
+      article: toObservable(this._articleDetailsStore.articleDetails),
```
Comment (DDonochVA) [inline libs/blog/shared/feature-seo/src/lib/seo/seo.service.ts:200]: We should avoid this. seo service is supposed to be a shared library that has no domain-related dependencies. Consider exposing methods that would be used inside the `ArticleDetailsStore`

File: libs/blog/shared/feature-seo/project.json [hunk 10]
```diff
@@ -0,0 +1,20 @@
+{
```
Comment (DDonochVA) [inline libs/blog/shared/feature-seo/project.json:1]: I think it should be an `util` library rather than `feature`

File: libs/blog/shared/feature-seo/src/lib/seo/seo.service.ts (hunk 8 above)
Comment (marcin-hoa) [reply libs/blog/shared/feature-seo/src/lib/seo/seo.service.ts]: @DDonochVA 
I only left `setMeta()` as public method which is bein called in `ArticleDetailsStore`.

`setTitle` is still being set directly in the `SeoService`, which is listenning to router events. This time it is only looking for `route?.data` - which is not set in Article Details Page. Let me know what you think

File: libs/blog/shared/feature-seo/src/lib/seo/seo.service.ts (hunk 9 above)
Comment (marcin-hoa) [reply libs/blog/shared/feature-seo/src/lib/seo/seo.service.ts]: @DDonochVA 
Look in comment above.

File: libs/blog/shared/feature-seo/src/lib/seo/seo.service.ts (hunk 6 above)
Comment (marcin-hoa) [reply libs/blog/shared/feature-seo/src/lib/seo/seo.service.ts]: Done

File: libs/blog/layouts/ui-navigation/src/lib/navigation/navigation.component.html [hunk 11]
```diff
@@ -11,6 +11,7 @@
       [routerLinkActive]="'text-al-pink'"
       [routerLink]="item.link | localize"
       [ariaLabel]="item.ariaLabel"
+      (click)="emitNavigated()"
```
Comment (mateuszbasinski) [inline libs/blog/layouts/ui-navigation/src/lib/navigation/navigation.component.html:14]:
```suggestion
      (click)="navigated.emit()"
```

File: libs/blog/shared/util-seo/project.json [hunk 12]
```diff
@@ -0,0 +1,20 @@
+{
+  "name": "article-seo",
```
Comment (mateuszbasinski) [inline libs/blog/shared/util-seo/project.json:2]: 👆

File: libs/blog/shell/feature-shell-web/src/lib/root-shell.component.ts [hunk 13]
```diff
@@ -37,6 +38,7 @@ import { SearchComponent } from '@angular-love/blog/search/feature-search';
   ],
 })
 export class RootShellComponent {
+  private seoService = inject(SeoService);
```
Comment (mateuszbasinski) [inline libs/blog/shell/feature-shell-web/src/lib/root-shell.component.ts:41]: is it needed here? If so, add _ and move to the bottom

File: libs/blog/shared/util-seo/src/lib/services/seo.service.ts [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+  private readonly _title = inject(Title);
+  private readonly _meta = inject(Meta);
+  private readonly _seoConfig = inject(SEO_CONFIG);
+
+  constructor() {
+    this.subscribe();
+  }
+
+  setMeta(seoData: SeoMetaData | undefined): void {
+    if (!seoData) {
```
Comment (DamianBrzezinskiHoA) [inline libs/blog/shared/util-seo/src/lib/services/seo.service.ts:26]: I'd consider using `switch`

File: libs/blog/layouts/ui-layouts/src/lib/header/header.component.html [hunk 15]
```diff
@@ -56,7 +56,11 @@
   class="bg-al-background fixed -top-[100%] left-0 z-50 flex h-full w-full flex-row items-start justify-between p-6 transition-transform duration-300 lg:hidden"
   [ngClass]="{ 'translate-y-[100%]': showNav() }"
 >
-  <al-navigation class="pl-24 pt-24" layout="vertical" />
+  <al-navigation
+    class="pl-24 pt-24"
+    (navigated)="toggleNav()"
```
[the remark this answers is not in the document]
Comment (marcin-hoa) [reply libs/blog/layouts/ui-layouts/src/lib/header/header.component.html:61]: Prettier formatting moves it automatically in the middle

File: libs/blog/shared/util-seo/src/lib/services/seo.service.ts (hunk 14 above)
Comment (marcin-hoa) [reply libs/blog/shared/util-seo/src/lib/services/seo.service.ts:26]: We have to check ALL props and set them. `switch-case` would only set one of the props.

File: libs/blog/articles/data-access/src/lib/state/article-details.store.ts [hunk 16]
```diff
@@ -28,6 +29,7 @@ export const ArticleDetailsStore = signalStore(
   withCallState('fetch article details'),
   withMethods(({ ...store }) => {
     const articlesService = inject(ArticlesService);
+    const seoService = inject(SeoService);
```
Comment (DDonochVA) [inline libs/blog/articles/data-access/src/lib/state/article-details.store.ts:32]: instead of this can you create signal store feature?
I'd like this to be hidden behind an abstraction, in case we would extend the seo.
```ts
export function withSeo() {
  return signalStoreFeature(
    ...
  );
}
``` 
^ and it would only have `withMethods`.

File: libs/blog/shell/feature-shell-web/src/lib/root-shell.component.ts [hunk 17]
```diff
@@ -39,6 +40,7 @@ import { SearchComponent } from '@angular-love/blog/search/feature-search';
 export class RootShellComponent {
   readonly translocoService = inject(TranslocoService);
   readonly localizeRouterService = inject(LocalizeRouterService);
+  private readonly _seoService = inject(SeoService);
```
[the remark this answers is not in the document]
Comment (DDonochVA) [reply libs/blog/shell/feature-shell-web/src/lib/root-shell.component.ts]:
```ts
{
   provide: APP_INITIALIZER,
   useFactory: () => {
     inject(SeoService).init()
   },
   multi: true
  }
```

**Review by DDonochVA — APPROVED** [review]
